### ISSUE TITLE
DI\object()->method() does not support calling same method twice

### ISSUE THREAD

**Issue by DizzyC** [mnapoli/PHP-DI#193]
This doesn't work:

```php
'My\AddNumbers' => DI\object()
        ->method('addNumber', 1)
        ->method('addNumber', 2),
```

Only `->method('addNumber', 2)` is taken into account.

The method `ClassDefinitionHelper::method()` ([code here](https://github.com/mnapoli/PHP-DI/blob/master/src/DI/Definition/Helper/ClassDefinitionHelper.php#L156)) uses the method name as a key for the [array](https://github.com/mnapoli/PHP-DI/blob/master/src/DI/Definition/Helper/ClassDefinitionHelper.php#L55) that holds methods that should be executed on the constructed object but if you try to call the same method twice it will obviously not work as the array key will be overridden, always retaining only the last method call.

**Comment by DHager** [comment]
Just chiming in: I got surprised by this today. I think it may be worth a note in the documentation, e.g.: "Multiple uses of the same setter are not supported."

I think the *reason* it was surprising is that the fluent/chained-method-call API suggests there's nothing wrong with using a bunch of them  in a big chain.

**Comment by mnapoli** [comment]
@DHager  thanks for the feedback. Yep I don't really like that situation too. I'm planning to improve that in 5.0 (currently working on it) but it will mean a bit of BC breaks.

**Comment by DHager** [comment]
One idea for a 4.0 workaround: Something like `ClassDefinitionHelper::afterward($func)` which runs  *after* everything else (instantiation, other setter injection) occurs. Building on a documentation example, suppose you want to invoke `setFoo` multiple times with multiple arguments?

    
    'My\OtherClass' => DI\object()
        ->scope(Scope::PROTOTYPE())
        ->constructor(DI\link('db.host'), DI\link('db.port'))
        ->method('setFoo2', DI\link('My\Foo1'), DI\link('My\Foo2'))
        ->property('bar', 'My\Bar')
        ->afterward(function(Container $c, $obj){
            $obj->setFoo($c->get("ThingOne"));
            $obj->setFoo($c->get("ThingTwo"));
            $obj->setFoo($c->get("ThingThree"));
        }),
    'My\ThingOne' => DI\object(),
    'My\ThingTwo' => DI\object(),
    'My\ThingThree' => DI\object(),
       

That way all the problematic setters can be tossed in there, without sacrificing the normal ways of dealing with constructor-injection, properties, etc.

**Comment by mnapoli** [comment]
The best workaround in 4.* is to use `DI\factory` though because it let you do all that already.

**Comment by DHager** [comment]
Doesn't it force you to give up on the constructor-argument autowiring?

**Comment by mnapoli** [comment]
Yes using a closure for that will force you to take care of instantiating the object entirely. Usually I need this pretty rarely so it's kind of acceptable (in my experience).

**Comment by mnapoli** [comment]
Fixed in 5.0 (to be released in the next months)
